### PR TITLE
Add additional strategy profitability statistic

### PR DESCRIPTION
**Pull request by m-root**
Closes https://github.com/tradingstrategy-ai/trade-executor/issues/31

I have made an update on strategy statistics update. Kindly have a look at it.

### CODE REVIEW

File: tradeexecutor/analysis/trade_analyser.py [hunk 1]
```diff
@@ -332,14 +329,27 @@ class TradeSummary:
...
     ))
 
+    average_winning_trade_profit_pc: float
+    average_losing_trade_loss_pc: float
+    biggest_winning_trade_pc: float
+    biggest_losing_trade_pc: float
+    average_duration_of_winning_trades: datetime.timedelta = field(metadata=config(
+        encoder=json_encode_timedelta,
+        decoder=json_decode_timedelta,
+    ))
```
Comment (plamut) [inline tradeexecutor/analysis/trade_analyser.py:347]: Curious, what info does the "duration" give us here? I know it's in the specs, but wouldn't it make more sense to track the maximum length of successive winning trades?

Asking because II can, for example, only make one trade a day, and with two successful trades the duration would be 24 hours. On the other hand, if I trade every hour and make 8 successive win trades, the duration of the streak would "only" be 8 hours, which is much less than 24 hours.

In other words, this value is heavily dominated by the frequency of the trades, meaning it's perhaps not all that informative? @miohtama can you give your thoughts here?

File: tradeexecutor/analysis/trade_analyser.py [hunk 2]
```diff
@@ -433,6 +451,11 @@ def calculate_summary_statistics(self) -> TradeSummary:
 
         duration = datetime.timedelta(0)
 
+        average_winning_trade = []
+        average_losing_trade = []
+        average_winning_trade_duration = []
+        average_losing_trade_duration = []
```
Comment (plamut) [inline tradeexecutor/analysis/trade_analyser.py:457]: Since these are all lists containing data points, would it make more sense to name them differently?

Because they do not actually represent the "average winning trade" and "average duration", this is computed later based on the data points in these lists.

File: tradeexecutor/analysis/trade_analyser.py [hunk 3]
```diff
@@ -451,14 +474,34 @@ def calculate_summary_statistics(self) -> TradeSummary:
...
+                average_losing_trade_duration.append(position.duration)
+
             else:
                 # Any profit exactly balances out loss in slippage and commission
                 zero_loss += 1
 
             profit += position.realised_profit
 
+        average_winning_trade_profit_pc = np.mean(average_winning_trade)
+        average_losing_trade_loss_pc = np.mean(average_losing_trade)
```
Comment (plamut) [inline tradeexecutor/analysis/trade_analyser.py:492]: Sanity check - do we really want to simply average the raw percentage values, or would it be better to use a weighted average taking trade sizes into account?

For example, if I risk 1 USD to buy an asset and then sell it for double the price (for 2 USD), II made a profit of 100%. Then in the next trade I risk 1000 USD and close the position for 1010 USD, realizing a modest profit of 1%.

According to the calculation, I made +55.5% per trade on average, but is this really the figure we want to report?
Because in reality I only made close to 1 with trading, because that 1000 USD trade dominates the other micro trade.

Perhaps a weighted average would be more informative?

File: tradeexecutor/analysis/trade_analyser.py [hunk 4]
```diff
@@ -451,14 +474,34 @@ def calculate_summary_statistics(self) -> TradeSummary:
...
+        if average_winning_trade:
+            biggest_winning_trade_pc = max(average_winning_trade)
+        else:
+            biggest_winning_trade_pc = np.nan
```
Comment (plamut) [inline tradeexecutor/analysis/trade_analyser.py:496]: Would it make more sense to just use `None` for unavailable stats?

**Review by plamut — COMMENTED** [review]
I have some questions about the definition/interpretation of some of the stats to make sure we indeed report what the traders expect.

We should also cover the changes with a test or two, there probably already exist tests for the trade stats?

And of course, we need to fix one of the tests that failed.

File: tradeexecutor/analysis/trade_analyser.py (hunk 1 above)
Comment (m-root) [reply tradeexecutor/analysis/trade_analyser.py:347]: As per the spec, it is the average duration of all winning trades combined. But it might be for consecutive successive trades.

File: tradeexecutor/analysis/trade_analyser.py (hunk 3 above)
Comment (m-root) [reply tradeexecutor/analysis/trade_analyser.py:492]: I agree with you, a weighted average would be more informative.

File: tradeexecutor/analysis/trade_analyser.py (hunk 4 above)
Comment (m-root) [reply tradeexecutor/analysis/trade_analyser.py:496]: We are returning the data to the notebook, not sure if `None` will be picked by the notebook. For notebook purposes that is why I am using `np.nan`

File: tradeexecutor/analysis/trade_analyser.py (hunk 2 above)
Comment (m-root) [reply tradeexecutor/analysis/trade_analyser.py:457]: Yes, I should name them differently.

File: tradeexecutor/analysis/trade_analyser.py (hunk 4 above)
Comment (hieuh25) [reply tradeexecutor/analysis/trade_analyser.py:496]: @m-root this isn't a correct assumption, the stats are used to displayed in notebook and will also be stored in state file (JSON) so you should find a way to make it suitable for both

**Review by hieuh25 — CHANGES_REQUESTED** [review]
Currently the PR fails the test suite, and also new test(s) for the new stats are missing, so it should be added.

File: tradeexecutor/analysis/trade_analyser.py (hunk 4 above)
Comment (plamut) [reply tradeexecutor/analysis/trade_analyser.py:496]: Fair point, let;s go with `None` then instead of a specific object from a library. `None` is so ubiquitous that "every" part of the ecosystem (libraries, renderers, serializers...) must know how to handle it.

File: tradeexecutor/analysis/trade_analyser.py (hunk 3 above)
Comment (plamut) [reply tradeexecutor/analysis/trade_analyser.py:492]: Just please make sure with @miohtama that this is indeed what is required by the spec. Perhaps the traders are used to a plain average instead of a weighted one.

**Review by plamut — COMMENTED** [review]
Thanks for the refactoring, the naming is less confusing now.

As discussed offline, we'll keep the current stats computation until somebody clearly requests a different definition (and computation).

As for the tests, do we already have a test that asserts that the computed averages and other values are indeed correct?
If not, we should add it (or add additional assertions to an existing test).

**Comment by miohtama** [comment]
Update documentation notebooks after merged to master.
